glib-macros/src/lib.rs

```diff
@@ -645,6 +645,59 @@ pub fn object_subclass(_attr: TokenStream, item: TokenStream) -> TokenStream {
     }
 }
 
+/// Macro for boilerplate of [`ObjectSubclass`] implementations specific to
+/// module types.
+///
+/// A module type must be explicitly registered when a module is loaded (see
+/// [`TypeModule`]).
+/// Therefore, unlike for non module types a module type can be registered
+/// several times.
+///
+/// A module type is never unregistered but is marked as unloaded by the glib
+/// type system when the module is unloaded (see [`TypeModuleExt::unuse`]).
+/// Therefore, a module type marked as unloaded by the glib type system must be
+/// registered again, when the module is reloaded.
+///
+/// This macro provides two behaviors when registering a module type:
+///
+/// By default the module type is registered when the module is loaded,
+/// ```ignore
+/// #[glib::module_object_subclass]
+/// impl ObjectSubclass for MyType { ... }
+/// ```
+///
+/// Optionally by setting the macro attribute to `lazy_registration = true` to
+/// postpone the registration of the module type on first use (when `type_()`
+/// is called for the first time), similarly to the [`macro@object_subclass`]
+/// macro.
+/// ```ignore
+/// #[glib::module_object_subclass(lazy_registration = true)]
+/// impl ObjectSubclass for MyType { ... }
+/// ```
+///
+/// [`ObjectSubclass`]: ../glib/subclass/types/trait.ObjectSubclass.html
+/// [`TypeModule`]: ../glib/gobject/auto/type_module/struct.TypeModule.html
+/// [`TypeModuleExt::unuse`]: ../glib/gobject/auto/type_module/trait.TypeModuleExt.html#method.unuse
+#[proc_macro_attribute]
+#[proc_macro_error]
+pub fn module_object_subclass(attr: TokenStream, item: TokenStream) -> TokenStream {
```

Comment:
Would it be possible to re-use the existing macro and just add configuration to it? `#[glib::object_subclass(dynamic)]` or so for example.

Or do you think having separate macros would be useful for better discoverability or so?

Reply:
Not sure. Either seems fine to me really :) Maybe @bilelmoussaoui has an opinion?

Reply:
There are trade-offs for both options but i think having a separate macro might make more sense as very few people would end up using it in general and it would make the documentation of object_subclass just more complex. On the other hand, we will have more macros and people will have a harder time to figure out when to use what

Reply:
Let's keep it a separate macro then :)